[Workers] Add tutorial: Live cursors & RPC with Next.js and Durable Objects (#18008)
cloudflare/cloudflare-docs: src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx

	class_name = "CursorSessions"

	[[migrations]]
	tag = "v1"
	new_classes = ["CursorSessions"]
	```
2. Initialize the Durable Object class in your `worker/src/index.ts` file:
	```ts title="worker/src/index.ts"
	import { DurableObject } from 'cloudflare:workers';

	export type WsMessage =
		| { type: "message"; data: string }
		| { type: "quit"; id: string }
		| { type: "join"; id: string }
		| { type: "move"; id: string; x: number; y: number }
		| { type: "get-cursors" }
		| { type: "get-cursors-response"; sessions: Session[] };

	export class CursorSessions extends DurableObject<Env> {
		constructor(ctx: DurableObjectState, env: Env) {}
		broadcast(message: WsMessage, self?: string) {}
		async webSocketMessage(ws: WebSocket, message: string) {}
		async webSocketClose(ws: WebSocket, code: number) {}
		closeSessions() {}
		async fetch(request: Request) {
			return new Response("Hello");
		}
	}

	export default {
		async fetch(request, env, ctx) {
			return new Response("Ok");
		},
	} satisfies ExportedHandler<Env>;
	```

	Now update `worker-configuration.d.ts` by running:

	```sh
	cd worker && npm run cf-typegen
	```

3. Update the Durable Object to manage WebSockets:

> 	- The main `fetch` handler routes requests with a `/ws` URL to the `CursorSessions` Durable Object where a WebSocket connection is stablished.

Reviewer: ```suggestion
	- The main `fetch` handler routes requests with a `/ws` URL to the `CursorSessions` Durable Object where a WebSocket connection is established.
```
Typo

	- The `CursorSessions` class manages WebSocket connections, session states, and broadcasts messages to other connected clients.

> 		- When a new WebSocket connection is established, the Durable Object broadcasts a `join` message to all connected clients; similaryly, a `quit` message is broadcasted when a client disconnects.

Reviewer: ```suggestion
		- When a new WebSocket connection is established, the Durable Object broadcasts a `join` message to all connected clients; similarly, a `quit` message is broadcast when a client disconnects.
```
Typo / Grammar

				},
			});
		},
	} satisfies ExportedHandler<Env>;

	```

4. Extend the `WorkerEntrypoint` class for RPC:
	:::note[Note]
	A service binding to `SessionsRPC` is used here because Durable Object RPC is not yet supported in multiple `wrangler dev` sessions. In this case, two `wrangler dev` sessions are used: one for the Next.js Worker and one for the Durable Object Worker. In production, however, Durable Object RPC is not an issue. For convenience in local development, a service binding is used instead of directly invoking the Durable Object RPC method.
	:::

	```ts title="worker/src/index.ts" ins={2,5-12} del={1}
	import { DurableObject } from 'cloudflare:workers';
	import { DurableObject, WorkerEntrypoint } from 'cloudflare:workers';
	// ... rest of the code

	export class SessionsRPC extends WorkerEntrypoint<Env> {
		async closeSessions() {
			const id = this.env.CURSOR_SESSIONS.idFromName("globalRoom");
			const stub = this.env.CURSOR_SESSIONS.get(id);
			// Invoking Durable Object RPC method. Same `wrangler dev` session.
			await stub.closeSessions();
		}
	}

	export default {
		async fetch(request, env, ctx) {
			if (request.url.match("/ws")) {
	// ...
	```

> 5. Run the Worker locally:

Author: it doesn't really matter where, the resulting address will be used in the next command. I make sure the readers leave the worker running for the local setup to work correctly, for instance the local websocket server

	<PackageManagers type="run" pkg="dev" frame="none" />

6. Add the Worker host as a Next.js public environment variable:
	```text title=".env.local" ins={1}
	NEXT_PUBLIC_WS_HOST=localhost:8787
	```

</Steps>

## 4. Build Next.js Worker functionality
<Steps>

> 1. In your `wrangler.toml` file, declare the external Durable Object binding and the Service binding to `SessionsRPC`:

Author: Here I'm updating the `next-rpc/wrangler.toml`, i'll be more explicit

				}}
			>
				<defs>
					<filter id="shadow" x="-40%" y="-40%" width="180%" height="180%">
						<feDropShadow dx="1" dy="1" stdDeviation="1.2" floodOpacity="0.5" />
					</filter>
				</defs>
				<g fill="none" transform="rotate(0 16 16)" filter="url(#shadow)">
					<path
						d="M12 24.4219V8.4069L23.591 20.0259H16.81l-.411.124z"
						fill="white"
					/>
					<path
						d="M21.0845 25.0962L17.4795 26.6312L12.7975 15.5422L16.4835 13.9892z"
						fill="white"
					/>
					<path
						d="M19.751 24.4155L17.907 25.1895L14.807 17.8155L16.648 17.04z"
						fill={color}
					/>
					<path
						d="M13 10.814V22.002L15.969 19.136l.428-.139h4.768z"
						fill={color}
					/>
				</g>
			</svg>
		);
	}
	```
	</Details>
	- It starts the WebSocket connection.
	- It handles 4 types of WebSocket messages:
		- `join`. Received when a new WebSocket connection is established.
		- `quit`. Received when a WebSocket connection is closed.
		- `move`. Received when a user's cursor moves.
		- `get-cursors-response`. Received	when a client sends a `get-cursors` message, which is triggered once the WebSocket connection is open.

> 	- It sends the user's cursor coordinates to the WebSocket server, which broadcasts them to all active WebSocket clients.

Author: Ok I added more details about how the client side logic works and updated the code to reflect a realistic scenario. the part responsible for animating the cursor, switching from css transitions to position interpolation.

5. Run Next.js locally:

	In your `package.json`, include a new script:

	```json title="package.json" ins={4}
		"scripts": {
			"dev": "next dev",
			"build": "next build",
			"build:worker": "cloudflare",
			"start": "next start",
			"lint": "next lint",
	```

	Build the app:
	<PackageManagers type="run" pkg="build:worker" frame="none" />

	Now, you can run Next.js development server:

> 	<PackageManagers type="run" pkg="dev" frame="none" />

Author: Did the build fail? I got eslint warnings, mainly caused by the code that highlights incoming/outgoing ws messages, which isn't directly relevant to the main content of the tutorial.

Yes, the durable object worker needs to be running. I'm adding a note

---
updated: 2024-11-05
difficulty: Intermediate
content_type: 📝 Tutorial
pcx_content_type: tutorial

> title: Live cursors & RPC with Next.js and Durable Objects

Reviewer 2: ```suggestion
title:  Display live cursors with RPC, Next.js and Durable Objects
```

Reviewer 2: I suggest making the title a present tense action to match our other tutorials. Maybe something like my suggested example, but feel free to change it.

- Once the WebSocket connection is established, the application sends a message to the WebSocket Durable Object to determine the current number of connected users.
- A user can close all active WebSocket connections via a Next.js server action that uses an RPC method.
- It handles WebSocket and mouse movement events to update the location of other users' cursors in the UI and to send updates about the user's own cursor, as well as join and leave WebSocket events.

![Animated gif of real-time Next.js app for visualizing live cursors](~/assets/images/workers/tutorials/live-cursors-nextjs/demo-live-cursors-nextjs-do.gif)

---

## Create a Next.js Workers Project

<Steps>

1. Run the following command to create your Next.js Worker named `next-rpc`:

	<PackageManagers
		type="create"
		pkg="cloudflare@latest next-rpc"
		args={"--framework=next --experimental"}
	/>

	<Render
		file="c3-post-run-steps"
		product="workers"
		params={{
			category: "web-framework",
			framework: "Next.js",
		}}
	/>

2. Change into your new directory:

	```sh
	cd next-rpc
	```

> 	Install [nanoid](https://www.npmjs.com/package/nanoid):

Reviewer 2: ```suggestion
3. Install [nanoid](https://www.npmjs.com/package/nanoid) so that string IDs can be generated for clients:
```

Reviewer 2: If the added description on how nanoid is used isn't accurate feel free to change the copy.

	<PackageManagers type="add" pkg="nanoid" frame="none" />

> 3. Define workspaces for each Worker:

Reviewer 2: ```suggestion
4. Define workspaces for each Worker:
```

	<Tabs> <TabItem label="npm">

	Update your `package.json` file.

> 	```json title="package.json" ins={14-17}

Reviewer 3: Hyperlint dev here, just to make sure I understand, this indentation is significant, right? It's a part of the step by step workflow and therefore indented?

Author: You’re right, it actually fails to build if the indentation is removed. There’s something about MDX here that might be causing it. I tried using spaces, but it looks like tabs are required

	{
		"name": "next-rpc",
		"version": "0.1.0",
		"private": true,
		"scripts": {
			"dev": "next dev",
			"build": "next build",
			"start": "next start",
			"lint": "next lint",

> 			"deploy": "cloudflare && wrangler deploy",

Author: The command `cloudflare` comes from the open nextjs package, I can't change it

			"preview": "cloudflare && wrangler dev",
			"cf-typegen": "wrangler types --env-interface CloudflareEnv env.d.ts"
		},
		"workspaces": [
			".",

> 			"worker"

Author: In this case `"worker"` is the name of a directory

	Create a new file `pnpm-workspace.yaml`.

	```yaml title="pnpm-workspace.yaml"
	packages:
		- "worker"
		- "."
	```
	</TabItem> </Tabs>

</Steps>

## Create a Durable Object Worker

This Worker will manage the Durable Object and also have internal API's
that will be made available to the Next.js Worker using a [`WorkerEntrypoint`](/workers/runtime-apis/bindings/service-bindings/rpc/) class.

<Steps>
1. Create another Worker named `worker` inside the Next.js directory:

	<PackageManagers
		type="create"
		pkg="cloudflare@latest"
		args={"worker"}
	/>

	<Render
		file="c3-post-run-steps"
		product="workers"
		params={{
			category: "hello-world",
			type: "Hello World Worker using Durable Objects",
			lang: "JavaScript / TypeScript",
		}}
	/>
</Steps>

## Setup Workers

### Worker project
<Steps>
1. In your `worker/wrangler.toml` file, update the Durable Object binding:
	```toml {4,5,9} title="worker/wrangler.toml"
	# ... Other wrangler configuration settings

	[[durable_objects.bindings]]
	name = "CURSOR_SESSIONS"
	class_name = "CursorSessions"

	[[migrations]]
	tag = "v1"
	new_classes = ["CursorSessions"]
	```
1. Initialize Durable Object class:
	```ts title="worker/src/index.ts"

> 	import { DurableObject } from 'cloudflare:workers';

Author: The indentation is necessary to prevent MDX from crashing and is part of the step-by-step workflow.
`'cloudflare:workers'` is a Cloudflare module, it is case-sensitive and cannot be changed.

1. Add the Worker host as a Next.js public environment variable:
	```text title=".env.local" ins={1}
	NEXT_PUBLIC_WS_HOST=localhost:8787
	```

</Steps>

### Next.js project
<Steps>
1. In your `wrangler.toml` file, declare the external Durable Object binding and the Service binding to `SessionRPC`:

	```toml title="wrangler.toml" ins={10-18}
	# ... rest of the configuration
	compatibility_flags = ["nodejs_compat"]

	# Minification helps to keep the Worker bundle size down and improve start up time.
	minify = true

	# Use the new Workers + Assets to host the static frontend files
	assets = { directory = ".worker-next/assets", binding = "ASSETS" }

	[[durable_objects.bindings]]
	name = "CURSOR_SESSIONS"
	class_name = "CursorSessions"
	script_name = "worker"

	[[services]]
	binding = "RPC_SERVICE"
	service = "worker"
	entrypoint = "SessionsRPC"
	```

1. Update your `env.d.ts` file for typesafety:
	```ts title="env.d.ts"
	interface CloudflareEnv {
		CURSOR_SESSIONS: DurableObjectNamespace<
			import("./worker/src/index").CursorSessions
		>;
		RPC_SERVICE: Service<import("./worker/src/index").SessionsRPC>;
		ASSETS: Fetcher;
	}
	```

> 1. Create server actions that make use of RPC methods:

Reviewer 2: ```suggestion
2. Create server actions that make use of RPC methods:
```

> 	The Service binding `RPC_SERVICE` is used instead of invoking Durable Object RPC methods because of the limitation mentioned in the note above.

Reviewer 2: I'd suggest adding some copy to describe how nanoid is used in the code block as well.

	import { getCloudflareContext } from "@opennextjs/cloudflare";
	import { Cursors } from "./cursor";
	import { nanoid } from "nanoid";

	export const dynamic = "force-dynamic";

	async function closeSessions() {
		"use server";
		const cf = await getCloudflareContext();
		await cf.env.RPC_SERVICE.closeSessions();

		// Note: Not supported in `wrangler dev`
		// const id = cf.env.CURSOR_SESSIONS.idFromName("globalRoom");
		// const stub = cf.env.CURSOR_SESSIONS.get(id);
		// await stub.closeSessions();
	}

	export default function Home() {
		const id = `ws_${nanoid(50)}`;
		return (
			<main className="flex min-h-screen flex-col items-center p-24 justify-center">
				<div className="border border-dashed w-full">
					<p className="pt-2 px-2">Server Actions</p>
					<div className="p-2">
						<form action={closeSessions}>
							<button className="border px-2 py-1">Close WebSockets</button>
						</form>
					</div>
				</div>
				<div className="border border-dashed w-full mt-2.5">
					<p className="py-2 px-2">Live Cursors</p>
					<div className="px-2 space-y-2">
						<Cursors id={id}></Cursors>
					</div>
				</div>
			</main>
		);
	}
	```

> 1. Create a client component to manage WebSocket and mouse movement events:

Reviewer 2: ```suggestion
3. Create a client component to manage WebSocket and mouse movement events:
```

						<feDropShadow dx="1" dy="1" stdDeviation="1.2" floodOpacity="0.5" />
					</filter>
				</defs>
				<g fill="none" transform="rotate(0 16 16)" filter="url(#shadow)">
					<path
						d="M12 24.4219V8.4069L23.591 20.0259H16.81l-.411.124z"
						fill="white"
					/>
					<path
						d="M21.0845 25.0962L17.4795 26.6312L12.7975 15.5422L16.4835 13.9892z"
						fill="white"
					/>
					<path
						d="M19.751 24.4155L17.907 25.1895L14.807 17.8155L16.648 17.04z"
						fill={color}
					/>
					<path
						d="M13 10.814V22.002L15.969 19.136l.428-.139h4.768z"
						fill={color}
					/>
				</g>
			</svg>
		);
	}
	```
	</Details>
	- It starts the WebSocket connection.
	- It handles 4 types of WebSocket messages:
		- `join`. Received when a new WebSocket connection is established.
		- `quit`. Received when a WebSocket connection is closed.
		- `move`. Received when a user's cursor moves.
		- `get-cursors-response`. Received	when a client sends a `get-cursors` message, which is triggered once the WebSocket connection is open.
	- It sends the user's cursor coordinates to the WebSocket server, which broadcasts them to all active WebSocket clients.

> 1. Run Next.js locally:

Reviewer 2: ```suggestion
4. Run Next.js locally:
```

	In your `package.json`, include a new script:

	```json title="package.json" ins={4}
		"scripts": {
			"dev": "next dev",
			"build": "next build",
			"build:worker": "cloudflare",
			"start": "next start",
			"lint": "next lint",
	```

	Build the app:
	<PackageManagers type="run" pkg="build:worker" frame="none" />

	Now, you can run Next.js development server:
	<PackageManagers type="run" pkg="dev" frame="none" />

</Steps>

## Deploy the project

<Steps>
1. Change into your Durable Object Worker directory:
	```sh
	cd worker
	```

	Deploy the Worker:
	<PackageManagers type="run" pkg="deploy" frame="none" />

	A URL is generated for the Worker, which will be used to update Next.js `NEXT_PUBLIC_WS_HOST` environment variable in your `.env.local` file.

	```txt title=".env.local" ins={2} del={1}
	NEXT_PUBLIC_WS_HOST=localhost:8787
	NEXT_PUBLIC_WS_HOST=<DEPLOYED-WORKER-HOST>
	```

> 1. Change into your root directory and deploy your Next.js app:

Reviewer 2: ```suggestion
2. Change into your root directory and deploy your Next.js app:
```

				// ...
			}
		}
	);
	```

	Each animated cursor is controlled by a `PerfectCursor` instance, which animates its position along a spline curve defined by the cursor's latest positions:
	```ts {9-11}
	// SvgCursor react component
	const refSvg = useRef<SVGSVGElement>(null);
	const animateCursor = useCallback((point: number[]) => {
		refSvg.current?.style.setProperty(
			"transform",
			`translate(${point[0]}px, ${point[1]}px)`,
		);
	}, []);
	const onPointMove = usePerfectCursor(animateCursor);
	// A `point` is added to the path whenever its vule updates;
	useLayoutEffect(() => onPointMove(point), [onPointMove, point]);
	// ...
	```

5. Run Next.js development server:
	:::note[Note]
	The Durable Object Worker must also be running.
	:::
	<PackageManagers type="run" pkg="dev" frame="none" />

6. Open the App in the browser.

</Steps>

## 5. Deploy the project

<Steps>
1. Change into your Durable Object Worker directory:
	```sh
	cd worker
	```

	Deploy the Worker:
	<PackageManagers type="run" pkg="deploy" frame="none" />

	A URL is generated for the Worker, which will be used to update Next.js `NEXT_PUBLIC_WS_HOST` environment variable in your `.env.local` file.

> 

Reviewer 4: Should I copy the whole URL? I copied the whole URL and ran into issues. Let's explicitly tell the reader to only paste the host and not the protocol (`https`)

	```txt title="next-rpc/.env.local" ins={2} del={1}
	NEXT_PUBLIC_WS_HOST=localhost:8787
	NEXT_PUBLIC_WS_HOST=<DEPLOYED-WORKER-HOST>
	```

> 2. Change into your root directory and deploy your Next.js app:

Reviewer 4: Do we still want to use Service bindings in the production? Earlier we told the user we are using it only for local development. So maybe we want to update the steps here?

Author: I think it can be an optional step. i'm adding a note mentioning the user can use the durable object rpc method instead of service bindings if they want.

---
updated: 2024-11-07
difficulty: Intermediate
content_type: 📝 Tutorial
pcx_content_type: tutorial
title: Build Live Cursors with Next.js, RPC and Durable Objects
products:
  - Workers

>   - Durable Objects

Reviewer 2: ```suggestion
  - Durable Objects
  - Developer Spotlight
spotlight:
  author: Ivan Buendia
  author_bio_link: https://www.linkedin.com/in/ibctx/
  author_bio_source: LinkedIn
```
